puppi/firmware/simple_puppi.cpp

```diff
@@ -0,0 +1,101 @@
+#include "simple_puppi.h"
+#include <cassert>
+#ifndef __SYNTHESIS__
+#include <cstdio>
+#endif
+
+typedef ap_uint<7> tk2em_dr_t;
+typedef ap_uint<8> weight_t;
+typedef ap_uint<10> tk2calo_dr_t;
+typedef ap_uint<10> em2calo_dr_t;
+typedef ap_uint<12> tk2calo_dq_t;
+typedef ap_uint<12> mu2trk_dr_t;
+
+// template<class data_T>
+// void lut_puppiweight_init3(data_T table_out[4096])
+// {
+//   for (int ii = 0; ii < 4096; ii++) {
+//     data_T real_val = weight_function_float( float(ii) );
+//     table_out[ii] = real_val;
+//   }
+// }
+
+weight_t puppiweight(weight_t iWeight){
+  weight_t _table[4096];
+  lut_puppiweight_init<weight_t,4096>(_table);
+  return _table[iWeight];
+}
+
+int dr2_int(etaphi_t eta1, etaphi_t phi1, etaphi_t eta2, etaphi_t phi2) {
+    etaphi_t deta = (eta1-eta2);
+    etaphi_t dphi = (phi1-phi2);
+    return deta*deta + dphi*dphi;
+}
+
+void compute_puppi_weight( ap_uint<32> eToAlpha, ap_uint<8> &curweight ){
+	
+	if (eToAlpha < 22026){
+		curweight = 0; // < e^10 where that is the median
+	} 
```

Comment:
if all these low values get truncated to zero, could we do the truncation at the beginning before doing the sum?
the total number of objects in the cone is at most O(20), so there's no way we can get to a very large eToAlpha by adding up very small values.
e.g. we could keep eToAlpha always bit-shifted right by 10 in the entire algorithm.

Reply:
agree, bitshift the sum